tsig.go

```diff
@@ -162,20 +162,30 @@ func tsigGenerateProvider(m *Msg, provider TsigProvider, requestMAC string, time
 	if err != nil {
 		return nil, "", err
 	}
-	buf, err := tsigBuffer(mbuf, rr, requestMAC, timersOnly)
-	if err != nil {
-		return nil, "", err
-	}
 
 	t := new(TSIG)
 	// Copy all TSIG fields except MAC and its size, which are filled using the computed digest.
 	*t = *rr
-	mac, err := provider.Generate(buf, rr)
-	if err != nil {
-		return nil, "", err
+
+	if rr.Error == 0 || rr.Error == RcodeBadTime {
+		// Only compute response MAC for non-errors and bad time errors
+		buf, err := tsigBuffer(mbuf, rr, requestMAC, timersOnly)
+		if err != nil {
+			return nil, "", err
+		}
+
+		mac, err := provider.Generate(buf, rr)
+		if err != nil {
+			return nil, "", err
+		}
+		t.MAC = hex.EncodeToString(mac)
+		t.MACSize = uint16(len(t.MAC) / 2) // Size is half!
+	} else {
+		// for TSIG errors (except bad time errors), respond with an empty MAC
```

Comment:
this can probably be pulled out of the if-statement and set unconditionally above it

Reply:
When testing this I tripped over a bug, which would require us to copy the TSIG struct twice for non-errors if we set unconditionally above it. I think it works better with the `else`... but if you're really opposed to the else, I can look for a way yo refactor the immediate code to accommodate it.

Reply:
> I can look for a way yo refactor the immediate code to accommodate 

Done. The `else` is gone.